src/main/java/io/aiven/kafka/auth/AivenAclAuthorizerV2.java

```diff
@@ -0,0 +1,224 @@
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+
+import io.aiven.kafka.auth.audit.AuditorAPI;
+import io.aiven.kafka.auth.json.AivenAcl;
+import io.aiven.kafka.auth.json.reader.AclJsonReader;
+import io.aiven.kafka.auth.json.reader.JsonReaderException;
+
+import kafka.network.RequestChannel.Session;
+import kafka.security.auth.Acl;
+import kafka.security.auth.Authorizer;
+import kafka.security.auth.Operation;
+import kafka.security.auth.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AivenAclAuthorizerV2 implements Authorizer {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AivenAclAuthorizerV2.class);
+    private File configFile;
+    private AuditorAPI auditor;
+    private boolean logDenials;
+    private final ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+    private WatchService watchService;
+    private final AtomicReference<VerdictCache> cacheReference = new AtomicReference<>();
+
+    public AivenAclAuthorizerV2() {
+    }
+
+    @Override
+    public void configure(final java.util.Map<String, ?> configs) {
+        final AivenAclAuthorizerConfig config = new AivenAclAuthorizerConfig(configs);
+
+        auditor = config.getAuditor();
+        logDenials = config.logDenials();
+
+        configFile = config.getConfigFile();
+        final AclJsonReader jsonReader = new AclJsonReader(configFile.toPath());
+        cacheReference.set(VerdictCache.create(loadAcls(jsonReader)));
+        final AtomicReference<WatchKey> watchKeyReference = new AtomicReference<>(subscribeToAclChanges(configFile));
+        scheduledExecutorService.scheduleWithFixedDelay(() -> {
+            final WatchKey watchKey = watchKeyReference.get();
+            if (watchKey != null) {
+                final List<WatchEvent<?>> watchEvents = watchKey.pollEvents();
+
+                watchEvents.stream().filter(watchEvent -> {
+                    @SuppressWarnings("unchecked")
+                    final Path path = ((WatchEvent<Path>) watchEvent).context();
+                    return path.endsWith(configFile.getName());
+                }).findFirst().ifPresent(watchEvent -> {
+                    LOGGER.info("{}: {}, Modified: {}",
+                            watchEvent.kind(), watchEvent.context(), configFile.lastModified());
+                    cacheReference.set(VerdictCache.create(loadAcls(jsonReader)));
+                });
+                if (!watchKey.reset()) {
+                    watchKeyReference.compareAndSet(watchKey, null);
```

Comment:
Maybe? Saving one iteration.
```suggestion
                    watchKeyReference.compareAndSet(watchKey, subscribeToAclChanges(configFile));
```